src/rust/src/lib.rs

```diff
@@ -1,28 +1,53 @@
+use async_tiff::{decoder::DecoderRegistry, TIFF};
 use extendr_api::prelude::*;
-use async_tiff::TIFF;
 //use async_tiff::decoder::DecoderRegistry;
-use async_tiff::reader::ObjectReader;
+use async_tiff::reader::{AsyncFileReader, ObjectReader};
+use tokio::runtime::{Builder, Runtime};
 
 //use std::io::BufReader;
-use std::sync::Arc;
+use std::{
+    cell::OnceCell,
+    sync::{Arc, OnceLock},
+};
 //use super::*;
 use object_store::local::LocalFileSystem;
 //use tiff::decoder::{DecodingResult, Limits};
 
+static TOKIO_RUNTIME: OnceLock<Runtime> = OnceLock::new();
+
+// Helper function to get a tokio runtime
+fn get_rt() -> &'static Runtime {
+    TOKIO_RUNTIME.get_or_init(|| {
+        Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .expect("Failed to create tokio runtime")
+    })
+}
 
 /// Return string `"Hello world!"` to R.
 /// @export
 #[extendr]
-fn rusty(file: String) -> &'static str {
-        let dir = "/tiff/";
-        let path = object_store::path::Path::parse(file).unwrap();
-        let store = Arc::new(LocalFileSystem::new_with_prefix(dir).unwrap());
-        let reader = Arc::new(ObjectReader::new(store, path));
-
-        // we can't .await.unwrap MDS
-        //let cog_reader = TIFF::try_open(reader.clone()).await.unwrap();
+fn rusty(file: String) -> Raw {
+    let dir = "/tiff/";
+    let path = object_store::path::Path::parse(file).unwrap();
+    let store = Arc::new(LocalFileSystem::new_with_prefix(dir).unwrap());
+    let reader = Arc::new(ObjectReader::new(store, path));
+    let rt = get_rt();
 
-    dir
+    let cog_reader = rt.block_on(async {
+        TIFF::try_open(Box::new(reader.as_ref().clone()))
+            .await
+            .unwrap()
+    });
```

Comment:
I think the example code you have is a smidge out of date. I modified it a bit to get it to work here.